### DISCUSSION TITLE
[Funnels] Is it possible to build a double opt-in flow with funnels?

### DISCUSSION THREAD

**Discussion by alexej-d** [dachcom-digital/pimcore-formbuilder#424] [Q&A] [answered]
I am trying to create a double opt-in flow with funnels, but the funnel functionality just isn't very well documented. I would just need to send a double opt-in email at the beginning of the funnel and once the user clicks the opt-in link in the mail the next actions / output workflows may be triggered. Any ideas or examples?

**Comment by solverat** [comment]
Hey @alexej-d, come on, the funnel feature is VERY well documented: https://github.com/dachcom-digital/pimcore-formbuilder/blob/master/docs/OutputWorkflow/40_Funnels.md! :)

You don't need funnels to achieve this. There are two ways:

### Event Listener 
- Add two mail channels (the second one is your double opt-in mail)
- Listen to `OUTPUT_WORKFLOW_GUARD_SUBJECT_PRE_DISPATCH`
- If the event holds your double opt-in mail and the form data does not provide the required checkbox, you should set `$event->shouldSuspend()` which silently skips the channel

### Conditional Logic
A bit more manual work but you don't have to write code for this:
- Create two output workflows (the first includes the default email channel, the second one the default email and the double opt-in mail channel)
- Create a conditional logic block ("If value checkbox is checked, action ⇾ switch output workflow")

***

For the sake of completeness, there's even a better way to achieve this, but still not implemeted: https://github.com/dachcom-digital/pimcore-formbuilder/issues/304

**Reply by alexej-d** [reply]
@solverat haha, sorry, let me rephrase that: The funnel functionality's documentation does not compute in my brain 😭 

I am not sure why you speak about a checkbox in your examples, maybe we have a misunderstanding. The way my double opt-in workflow should work is follows:
 user submits a form -> user receives a double opt-in mail with an individual link -> clicking the link "releases" the other mail workflow where the user's personal data is sent to whoever

If the link is not clicked for like 2 weeks it expires.

**Reply by solverat** [reply, marked as answer]
Ah, now I got you. Sadly, there is no solution out of the box. Even funnels aren't the right thing here, because they always start immediately after a form has been dispatched. Also the individual link should be stored somewhere, right? Otherwise, we can't reassign the user.

It's a bit out of context, but I think your workflow - with formbuilder maximum involved :) - should be:

- Create a form
- Add a object channel and a mail channel
- After form gets submitted, populate an object (you may want to use a [custom resolver](https://github.com/dachcom-digital/pimcore-formbuilder/blob/master/docs/OutputWorkflow/11_ObjectChannel.md#ii-create-a-custom-resolver) for that) which also holds a unique link
- Generate a Route/Controller/Action and listen to the unique link ⇾ resolves object ⇾ outdate it if older than two weeks, or...
- ...then [manually](https://github.com/dachcom-digital/pimcore-formbuilder/blob/master/docs/0_Usage.md#usage-iii-controller) render a second form within this action, which just - i don't know :) - holds some fields a user has to fill up to complete the journey 
- This second form then triggers another workflow after the user hits the submit button
  - btw. this is the moment, a funnel could rise up :).
- Done

That's a rough walkthrough, which I would choose to get the most out of form builder, but I'm afraid, it is not possible to do it without any coding. 

Hope this helps!

**Reply by alexej-d** [reply]
@solverat thanks, that might be a nice way to go! Right now, I extend the email output workflow with another mail template field for the double opt-in email and dispatch two messages (via Symfony's message component). One message creates the opt-in entity, the link and the opt-in email, the other is being suspended until the link is clicked or expired. Still, I will mark your comment as the answer :)
